### ISSUE TITLE
Enable mail sending for new CRI

### ISSUE THREAD

**Issue by damirius** [Puzzlout/TrekToursApi#24]
Just need to enable again usage of swift mailer for sending CRI notification to the admin and cc to the sender (if has_sent_copy_to_client is set). Also need to write tests for it.

**Comment by WebDevJL** [comment]
@damirius Good initiative. You can use the account the parameters.yml.dist, as you probably guessed.

**Comment by damirius** [comment]
@WebDevJL 

This is done. Email should be sent to the admin_mail_address specified in parameters also it should be sent from that email. So admin mail address should have same host we are sending our mails from to prevent that it goes to spam. Client will be set as CC if he checked send me an email copy checkbox on the contact form. If admin_mail_address is set to ~ which is null, no emails will be sent.
As for the tests, I'm creating mockup object of switft mailer so don't worry about sending emails while testing, same goes for travis tests of course.
Everything is on this feature_customerinforequest branch

**Comment by WebDevJL** [comment]
@damirius I'd like a few changes to that functionality if you don't mind:

The first email sent should be to contact@puzzlout.com from noreply@puzzlout.com.
The subject "You received a message from a customer" should be translated based on the admin language. How much work is that to do?

The second email, if the customer checked to receive a copy, is sent to the customer from noreply@puzzlout.com. 
The subject "A copy of your request sent to TrekTours" should be translated based on the admin language. 

Thanks.

**Comment by damirius** [comment]
@WebDevJL 

Hm, changes in the functionality shouldn't be to hard to implement. So we should sent two emails instead of one with cc? Is there any difference between two emails apart from subject?

Translations are our biggest problem here. Existing translations are on flyer website and they are managed there, since translations are linked to the flyer app itself and flyers design.
Also what do you mean by "translated based on the admin language"? Where do we fetch admin language from? New CRIs are made from contact page on flyer website, but they also can be made with other apps through the API.

**Comment by WebDevJL** [comment]
> @damirius said:
> Hm, changes in the functionality shouldn't be to hard to implement. So we should sent two emails instead of one with cc? Is there any difference between two emails apart from subject?

yes, two emails sent to different recipients. 

The body of the email to the admin should be:

```
Hello,

You received a message from {{ name of customer }}:

{{ customer message }}
```

The body of the email to the client if he checked the box to receive a copy should be:

```
Hello {{ name of customer }},

You have sent the following message:

{{ customer message }}
```

**Comment by WebDevJL** [comment]
@damirius I'm thinking about the translation. Give me a moment to compose my answer

**Comment by WebDevJL** [comment]
> @damirius:
> Translations are our biggest problem here. Existing translations are on flyer website and they are managed there, since translations are linked to the flyer app itself and flyers design.
> Also what do you mean by "translated based on the admin language"? Where do we fetch admin language from? New CRIs are made from contact page on flyer website, but they also can be made with other apps through the API.

I will ask to Stéphane about this. 
Now, in the meantime, let's say that we have a setting that defines the default language of the site admin. 

For the email to the admin, can we retrieve the email body and subject from the Flyer's translations based on the setting on pass it on to the API? This means to modify the API that adds a CRI to include the body and subject. Is that ok?

For the email copy sent to the customer, same thing but the language is the current language selected. Translations would also come from the translation base of the Flyer.

Let's me know what you think about that and we will create an issue separate from this one. Thanks.

**Comment by damirius** [comment]
@WebDevJL 

OK, so we have some language set in parameters or something like that for the admin part. We send subject and whole email message translated to that language for admin email. Then we fetch current language client is using and translate subject and message for him as well.
We send both subjects and email bodies with other details to the api which will just send those messages like that?
We can do it like that. It's little weird to send whole message through the API, but I can't think of any other solution.
Only thing I can think about is that we send only language shorthand for the client and the admin ('en', 'fr') to the API and let the API translate it. Bad thing about that is that we would need to create translation yaml files which will contain those translations and we couldn't change them through that admin section for translations on the flyer. If we would ever want to change them we would need to edit translation files on the API.

**Comment by damirius** [comment]
@WebDevJL 

I thought little about this, it's quite peculiar problem we are facing here.
First off I want to say that I really want to follow RESTful standards as much as we can. I know that we are currently using API just for one application (flyer), but we should work on it like it serves multiple apps at once even if we stay on one app for long time. Since we will upgrade and work more on that API I don't want to create solutions now that turn to be the source of the problems later on.
Let's look at our choice of keeping our translations on the Application (flyer) side. All translations there are application specific, let's say that we create mobile app. That app will have different menus, pages, links and so on, there will be almost no use of translations we made for the flyer application. Plus there is technical standpoint where we would need to make costly calls to the API so we could fetch translations for each page we need. So we would need to make some kind of caching system there, then we would need to create some system of clearing cache on translation updates, then we will need to clear the cache only for specific translations and so on. It's a slippery slope like the most things built like this.
All that said let's move on our problem with emails, I'll try to provide some solutions and explanations for each one, but I will need your input to see what exactly do we need.
Core of the problem is always similar. What's the job of the API and what's the job of the application, in our case, flyer website.
**Solutions:**
-  **Send email contents with the CRI call**

This is the solution we talked about. We create email body templates for both admin notification email and client notification email in our app (Flyer). We translate the client email template based on the language they chosen on the website and we translate the admin email template based on preset admin language configuration (most likely set in the parameters.yml file). Then we send both of those email bodies together with translated email subjects to the API with other details (name, email, phone number etc.)
**For:** We can easily manage translations from our existing translation panel and change the template and subject translation on any language, even on some new languages if we want, from our flyer website. We can even change whole templates on each application. API doesn't need to know anything about it, it just needs to send those emails.
**Against:** We are sending whole emails through the API call. Since clients can be anonymous we don't have authorization for that call, unlike for the other calls to the API. Anyone can just change the email message even for the admin. Also they can just enter some email they want to send spam to and change the subject and the body of the message to send it. Each app that is using the API is required to send both emails through the call and that can't be easily changed.
**Complexity of implementation:** 6/10 - We need to create templates on the Flyer project, change the call on the API side with additional validations for the messages and subjects
- **Send only language to translate to API**

We send only admin language and client language to the API. Instead of translating everything on the application side we send shorthands of admin email language and of client email language. There are two ways of implementing this and they are similar enough that I bundled them in this solution.

_First one_ is to set admin language on the API side, we already have "admin address" there (we need noreply address too), we just need to add "admin_email_language" there too which we will fetch on create CRI action and we can send client email language from the application (flyer) through "Accept-Language" header which is often used for this things and use it to translate the client email subject and body. Accept-Language header only accepts one language, so we can send one shorthand for _either_ admin email _or_ client email.

_Second One_ is to send both admin and email language through our application. We set "admin_email_language" in parameters like we did in first solution and send it together with "client_email_language" as POST parameters through the call to the API. It's not so standard since those two parameters have nothing to do with CRI resource, but it's not such big problem.

For both solutions we take those language shorthands and translate both subject and the body for both emails based on translations _stored on the API side_ in yml files. So we would have to create manually translation files, for example `email.en.yml` and create keys inside `email_client_body`, `email_client_subject`, `email_admin_body`, `email_admin_subject` with translations.
**For:** There's no security issue here, only thing that can be changed is the language. If language isn't supported we fallback to default language. No need for sending subject and body for both emails from the app, API is doing all the work. App only needs to send client language and/or admin language depending on solution chosen. More standard solution.
**Against:** We can't change translations easily. If we want to change the translations we would need to edit yml files located on the API side which should be done manually. Same goes for templates, they would need to be changed manually. We serve same templates for each application and only in those languages set on API.
**Complexity of implementation:** 5/10 - We need to implement translations on the API side and change the call to accept headers or new parameters, on the flyer we only need to send one or two language shorthands.
- **Send emails from the Application**

If admin/client notification is depending on the app that's using the API, this is the best solution. We remove email sending from the API and move it to the application. API will only create new CRI and store it in the DB. Applications should care about sending notifications if they want to on successful creation of CRI (on return code 201). When we get response from the API we create new emails for admin and/or client and send it to them translated. Translations are stored in application like in first solution, we just don't need to send them to the API we send them from the app. API will still be told if notifications are sent if needed through `has_sent_copy_to_client` flag or we can remove it completely if other applications don't need to know about this.
**For:** Applications are taking care of email notifications, we can send them if we wish or not send them at all. We can send them to the other admin email for each application (for example different domain) or even to multiple admins if we want from the same application, we can send notification to the client in one app and in the other we can turn that off completely. We can change templates and translations in each app so they are completely flexible. API will still serve it's purpose and provide essential data to all apps we need.
**Against:** Each app has to do this separately since API has nothing to do with it. API doesn't know contents of the emails and it doesn't care if we sent the email or not or to whom exactly (what address). We can only tell it that we sent the email copy to the client if other apps need to know this. We would need to implement specific solutions for each application for sending emails.
**Complexity of implementation:** 3/10 - Need to move email sending code to the flyer website from the API side and move email params from the config.
- **API managing email sending and templates**

This is most time consuming solution, but it's rather flexible and more elegant. If we need our API to manage all email transfers we can go with this. We create small database table which will contain email templates for each language we want. Then we make CRUD calls for templates (POST, PUT, GET, DELETE at least) for /templates. Making a POST request with _language_, _template_ and _key_ parameters will create new template translation. GET with _key_ and _language_ query params will display the template and DELETE with delete it. We can even add `user_id`/`app_id` field there so that each app admin has it's own templates if we want to separate templates for each application.
Making a POST call to create new CRI with language and app_id parameters will fetch template based on `app_id`, `language` and preset key (`email_admin_body`/`email_admin_subject`) then do the same for the client email. We can even create some settings table where we will store settings for each application (like flyer) for example `admin_email_language` or whatever we want. App will have to have some sort of admin panel for managing those templates (and settings if we want that).
**For:** Rather flexible solution on the long run where we implement best things from first two solutions. We are still managing templates and translations on the app side, but we let API send the emails without sending unnecessary stuff in the call. If we add `app_id` in the table we can use different templates/translations for each app. If we add settings table, we can preset admin_email_language without sending it every time through the call, and we can store different stuff there. **Most importantly** we can use this for different templates for different stuff later on, not just emails.
**Against:** VERY, very time consuming as you probably noticed from the explanation. It isn't just solution for emails it's for all templates/settings that are app specific, which is the good thing, but it's little too much if we are talking about this specific issue only.
**Complexity of implementation:** 10/10 - Rather big changes on the API side, at least 4 new calls for templates and at least 4+ for settings if we want that too, change of existing call also. New admin pages on the flyer (app) website which will take care of settings/translations or at least new code triggers if we are gonna hook templates into existing translation admin page. So something like first two solutions combined with additional stuff on the API and the Flyer. Honestly I have no problem with starting to implement this but it will take some time.

_Note that Complexity of implementation_ is relative to other solutions.

Thanks for reading this, I know it's a lot, but I think it's for the best that we lay out some details here before starting to do something like this. I know some things often look easy to implement and some others look too hard, so I just wanted to give as detailed preview of solutions as I could.

**Comment by WebDevJL** [comment]
@damirius

Thank you for your detailed explanation. This is good stuff.

Now about the solution to choose, I think that we could do the more
flexible, complete and future proof solution when we will develop the
storefront.
For the storefront, we will need that solution because we will have all
orders and account creation notifications to send to customers and
administrators and management of email templates and settings in general
will be critical.

**_For now, I don't think we have budget for a 10/10 solution.**_

Solution 3/10 will meet our budget for the flyer.
One thing that we could add are endpoints to notify the api that a CRI was
sent successfully to the administrator and, if applicable, to the customer.
This means to add two new properties to a CRI :
administrator_email_sent_date and customer_email_sent_date.
Why do we need this? To know that the emails are sent or not. The creation
of the CRI and the sending of the email will be done very to each other.

It also creates the question of processing emails notification not sent.
What if the smtp server is down? This will be another issue.

Let know if solution 3/10 can be with the small addition and if you find
the addition relevant. If so that will be separate from this issue.

Thank you.

## 

Jérémie Litzler

## 

Jérémie Litzler

**Comment by damirius** [comment]
@WebDevJL 

OK, I think that's best solution for now, maybe even better that 4th solution that works with the templates and what not.
We could add those two fields `administrator_email_sent_date` and `customer_email_sent_date` to the CRI table, they would default to null and on CRI creation they will be left as null. After we make a call from a flyer (or any other app) CRI will be created with those two fields left as nulls. We also have that `has_sent_copy_to_client` field, which should be maybe renamed to `send_copy_to_client` instead, but in any case it will help us to see if we need to send CRI copy to client or not at all. After that call is made we can send emails on successful entry from our application and if emails were sent correctly we can make separate PATCH call where we will update `administrator_email_sent_date` and `customer_email_sent_date` to current date.
Now if `administrator_email_sent_date` is not null we know that we sent the email to the admin. If `has_sent_copy_to_client` and `customer_email_sent_date` are both set we know that we sent the email to the customer/client. If `administrator_email_sent_date` is null we know that we had some problems with sending email to the admin. If `has_sent_copy_to_client` is set to 0 we know that we don't need to send email to client so we don't care for date field, if `has_sent_copy_to_client` is set to 1 but `customer_email_sent_date` is set to null we know that we also failed sending email to the customer. I know that most of the times if we fail sending one email we will fail sending other one too and vice versa, but just want to make sure everything is covered.
Now with that in mind we can for example mark somehow any records in the CRI admin section which have those fields set to null and add some button to resend the email(s). But since that part is for the storefront we can talk about it there. For now we should focus on the API part and using it in the flyer.

We can open up new issue on the flyer repo too, since we will need to implement most of the things there. After that I can use this issue to make changes to the API side.

**Comment by WebDevJL** [comment]
@damirius Go ahead with that. Open up the new issue. Thanks.
